universe/Condition.cpp

```diff
@@ -9956,6 +9965,249 @@ unsigned int Not::GetCheckSum() const {
+    // If no candidate matches A, B, or C, then nothing is matched.
+    //
+    // Not OrderedAlternativesOf [ A B C ] finds the topmost condition which has matches and then matches its non-matches.
+    // If any candidate matches A, then all candidates that do not match A are matched,
+    // or if no candidates match A but any candidate matches B, then all candidates that do not match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that do not match C are matched.
+    // If no candidate matches A, B, or C, then all candidates are matched.
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
+        // stop if the operand condition has at least one match
+        // Return non-matches if there are any
+        ObjectSet temp_non_matches;
+        temp_non_matches.reserve(non_matches.size());
+        // try matching operand conditions until one matches
+        bool found_sth = false;
+        for (auto& operand : m_operands) {
+            // move those non matching the current operand to temp_non_matches
+            //            operand->Eval(local_context, matches, temp_non_matches, MATCHES);
+            // move those matching the current non-negated operand from non_matches to temp_non_matches
+            operand->Eval(local_context, non_matches, temp_non_matches, MATCHES);
+            if (!non_matches.empty()) {
+                // i.e. the current operand should be used negated to move
+                found_sth = true;
+            }
+            // recreate previous state by moving back temp_non matches to non_matches
+            FCMoveContent(temp_non_matches, non_matches);
+
+            if (!found_sth) {
+                 // Also have to look at the other local candidates if there was a match
+                ObjectSet temp_matches;
+                temp_matches.reserve(matches.size());
+                operand->Eval(local_context, temp_matches, matches, NON_MATCHES);
+                if (!temp_matches.empty()) {
+                    // i.e. the current operand should be used negated to move
+                    found_sth = true;
+                }
+                // recreate previous state
+                FCMoveContent(temp_matches, matches);
+            }
+
+            if (found_sth) {
+                // descent into subcondition for NON_MATCHES
+                operand->Eval(local_context, matches, non_matches, NON_MATCHES);
+                break;
+            }
+            // try the next operand
+        }
+        // If NON_MATCHES could not find any matches it should match all
+        if (!found_sth)
+            FCMoveContent(matches, non_matches);
+    } else /*(search_domain == MATCHES)*/ {
+        // check all operand conditions on all objects in the matches set, collecting non matches temporarily
+        // stop if the operand condition matches at least one item
```

Comment:
// check each operand condition on objects in the input matches and non_matches sets, until an operand condition matches an object. if an operand condition is selected, apply it to the input matches set, moving non-matching candidates to non_matches. if no operand condition is selected, because no candidate is matched by any operand condition, then move all of the input matches into non_matches.